docs/reference/esql/functions/parameters/ip_prefix.asciidoc

```diff
@@ -0,0 +1,9 @@
+// This is generated by ESQL's AbstractFunctionTestCase. Do no edit it. See ../README.md for how to regenerate it.
+
+*Parameters*
+
+`ip`::
+IP address of type `ip` (both IPv4 and IPv6 are supported).
+
+`prefixLength`::
+Prefix length. If in the range (0, 32), the IP is treated as an IPv4 address. If in the range (32, 128), the IP is treated as an IPv6 address. If the prefix length is out of range, the function returns `null`.
```

Comment:
This is the edge case here, and I think @nik9000 had other ideas for this too. My thought here was:
1. Yes, this makes it impossible to prefix IPv6 by 24.
    - If the case appeared, my idea was to create a new `ip_prefix_v6()` function. This one would be the "Generic, IPv4 one, with IPv6 support"
2. That special treatment is related with how we store IPv4 in lucene. It's stored as 16 bytes ref (The same as IPv6), with the IPv4 stored in the last 4 bytes. It's possible to detect if it's a v4, as the first 10 bytes (Or something like that) are 0, plus 2 constant bytes (Or 4, whatever).

So I see multiple alternatives:
1. v4 + most v6 function (Current one)
2. Dynamically detect if the IP is v4 or v6. Apply one or other logic accordingly (The one you're proposing I think? Which looks good to me, should have chose this one tbh)
3. v4 function and null on v6 (Would require a second function for v6)

Going with 2 then? Want to confirm if we're all in the same page

Reply:
Yeah, the ipv6 handling is hard to wrap your head around either way. I'd told @ivancea to copy the behavior of the `ip_prefix` agg which does something like this. But, yeah, it's funky.

I wonder if we'd be better off just calling this one `ipv4_prefix` and documenting that it only allows (0, 32). The funky thing is that it's going to operate on ipv6 address too because that's how our `ip` field works. We embed ipv4 into ipv6 copying the behavior from some rfc. Presuming we just call it `ipv4_prefix`, the question I have is, "what happens when you send it an ipv6 address?" It could return `null` and warn - or it could slice wherever the slice would be to align with the rfc.

One advantage we have is that we can make an `is_v6` function for `ip` and users can decide on what to do when they bump into v6 with that.

Lookup at #99064 it looks like they have a v4 and v6 cidr. Maybe we allow that? `ip_prefix(ip_field, 24, 128)` and it's got the two choices built right in.

Reply:
> Maybe we allow that?

Maybe we just force that. Do the two argument version now and then wait until folks ask for a one-argument version and see what shape they mean. Maybe we'll even have named-args by then.

Reply:
Updated the PR to have 3 params:
- ip
- lengthPrefixV4
- lengthPrefixV6 (Optional, defaults to 128 to not mutate the v6 ips)